### PR TITLE
Major refactor #1

### PR DESCRIPTION
**Pull request by myabc**
#5
#12
#8

### CODE REVIEW

**Comment by apotonick** [comment]
Send it! :package:

**Comment by fran-worley** [comment]
:shipit: 👍

**Comment by myabc** [comment]
@apotonick @fran-worley thanks! :-) This PR is finally progressing. I'd appreciate some early feedback though on work to support [relationship object links and meta](http://jsonapi.org/format/#document-resource-object-relationships) - both on Ruby API design and implementation (will add some questions inline)

File: lib/roar/json/json_api/declarative.rb [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+          return super(name, &block) unless options[:toplevel]
+          for_collection.link(name, &block)
+        end
+
+        def meta(options = {}, &block)
+          return super(&block) unless options[:toplevel]
+          for_collection.meta(&block)
+        end
+
+        def relationship(options = {}, &block); end
```
Comment (myabc) [inline lib/roar/json/json_api/declarative.rb:23]: (no-op here)

File: lib/roar/json/json_api/declarative.rb [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+          resource_identifier_representer = Class.new(Roar::Decorator)
+          resource_identifier_representer.class_eval do
+            include Roar::JSON
+            include Roar::Hypermedia
+            include JSONAPI::Meta
+            extend JSONAPI::Declarative
+
+            def self.relationship_block
+              @relationship_block || -> {}
```
Comment (myabc) [inline lib/roar/json/json_api/declarative.rb]: @apotonick there is probably a much nicer way of doing this, but this is what I've come up with so far to support the following API:

```ruby
has_many :comments, class: Comment, populator: ::Representable::FindOrInstantiate do
  type :comments

  relationship do
    link(:self)     { "/articles/#{represented.id}/relationships/comments" }
    link(:related)  { "/articles/#{represented.id}/comments" }
  end

  # property :id
end
```

Comment (myabc) [reply lib/roar/json/json_api/declarative.rb]: (`relationship` is conceptually similar, but not exactly the same as the current `toplevel` option for links, etc. - hence different approach)

Comment (apotonick) [reply lib/roar/json/json_api/declarative.rb]: What's the difference here? Couldn't we do `relationship` with `link` instead?

Comment (myabc) [reply lib/roar/json/json_api/declarative.rb]: A JSON API `relationship` has special semantics, hence my preference for a `relationship` block. In fact, reading the spec again, it appears even empty to-one relationships can have links/meta associated with them.

However, let's suppose we want to implement this in a similar way to the way `toplevel: true` is implemented:

```ruby
  link(:self, :relationship: true)     { "/articles/#{represented.id}/relationships/comments" }
  link(:related, :relationship: true)  { "/articles/#{represented.id}/comments" }
  meta :relationship: true do
    property :count
  end
```

This is predicated on the `::for_collection` API:

```ruby
def link(name, options = {}, &block)
  return super(name, &block) unless options[:toplevel]
  for_collection.link(name, &block)
end
```

AFAICT there is no built-in mechanism for a [child representer](https://github.com/trailblazer/roar-jsonapi/pull/16/files#diff-5cf99150a8f672cf05cc67c7a054c120R80) to keep a reference to [its parent](https://github.com/trailblazer/roar-jsonapi/pull/16/files#diff-5cf99150a8f672cf05cc67c7a054c120R74).

```ruby
def link(name, options = {}, &block)
  return for_collection.link(name, &block) if options[:toplevel]
  return for_relationship.link(name, &block) if options[:relationship] # ref to parent
  super(name, &block)
end
```

Or am I missing something?

Comment (apotonick) [reply lib/roar/json/json_api/declarative.rb]: We can always simplify it later... :+1:

**Pull request merged**